Integration tests fail with nested jpaDerivedIdentifier

<!--
- Please follow the issue template below for bug reports.
- If you have a support request rather than a bug, please use [Stack Overflow](http://stackoverflow.com/questions/tagged/jhipster) with the JHipster tag.
- For bug reports it is mandatory to run the command `jhipster info` in your project's root folder, and paste the result here.
- Tickets opened without any of these pieces of information will be **closed** without any explanation.
-->

##### **Overview of the issue**
Integration tests fail when I define nested entities using jpaDerivedIdentifier
the integration tests fail for entities at or lower than grandchild level

Given the following entities:  
  - Animal
    - Fish
    - Bird
    - Mammal
      - Marsupial
      - Rodent
      - Cat
        - DomesticatedCat

And the following relationships:
relationship OneToOne {
  Fish{animal required} to Animal with jpaDerivedIdentifier
  Bird{animal required} to Animal with jpaDerivedIdentifier
  Mammal{animal required} to Animal with jpaDerivedIdentifier
    Marsupial{mammal required} to Mammal with jpaDerivedIdentifier
    Rodent{mammal required} to Mammal with jpaDerivedIdentifier
    Cat{mammal required} to Mammal with jpaDerivedIdentifier
      DomesticatedCat{cat required} to Cat with jpaDerivedIdentifier
}

generate the app
> jhipster jdl tree.jdl

test the app 
> mvn verify

The following errors are reported:

> [ERROR] Errors:
> [ERROR]   CatResourceIT.updateCatMapsIdAssociationWithNewId:150 » EntityExists A differe...
> [ERROR]   DomesticatedCatResourceIT.updateDomesticatedCatMapsIdAssociationWithNewId:155 » EntityExists
> [ERROR]   MarsupialResourceIT.updateMarsupialMapsIdAssociationWithNewId:150 » EntityExists
> [ERROR]   RodentResourceIT.updateRodentMapsIdAssociationWithNewId:150 » EntityExists A d...
> 

The tests for AnimalResourceIT, FishResourceIT, BirdResourceIT and MammalResourceIT all work,  
but anything lower than one level has a problem

<!-- Explain the bug, if an error is being thrown a stack trace helps -->

##### **Motivation for or Use Case**

<!-- Explain why this is a bug for you -->
This isn't a problem for the app  at runtime, it seems to be an integration test issue only.

##### **Reproduce the error**
<!-- For bug reports, an unambiguous set of steps to reproduce the error -->
See the Overview, jdl will be provided.

##### **Related issues**
<!-- Has a similar issue been reported before? Please search both closed & open issues -->


##### **Suggest a Fix**
<!-- If you can't fix the bug yourself, perhaps you can point to what might be
  causing the problem (line of code or commit) -->

##### **JHipster Version(s)**
<!--
Which version of JHipster are you using, is it a regression?
-->
7.5.0

##### **JHipster configuration**
<!--
To provide all information we need, you should run `jhipster info` in the project root folder, and
copy/paste the result here.
The `.yo-rc.json` file generated in the root folder is mandatory for bug reports. This will help us to replicate the scenario.
You should remove any sensitive information like the rememberMe key or the jwtSecretKey key.
--->
INFO! Using JHipster version installed locally in current project's node_modules
Welcome to the JHipster Information Sub-Generator

##### **JHipster Version(s)**

```
tree@0.0.1-SNAPSHOT C:\github\jhip-treebug
`-- generator-jhipster@7.5.0

```


##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**


<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "authenticationType": "jwt",
    "cacheProvider": "ehcache",
    "clientFramework": "angularX",
    "serverPort": "8080",
    "serviceDiscoveryType": false,
    "skipUserManagement": false,
    "withAdminUi": true,
    "baseName": "tree",
    "buildTool": "maven",
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "enableHibernateCache": true,
    "enableSwaggerCodegen": false,
    "enableTranslation": true,
    "jhiPrefix": "jhi",
    "languages": ["en"],
    "messageBroker": false,
    "prodDatabaseType": "postgresql",
    "searchEngine": false,
    "testFrameworks": [],
    "websocket": false,
    "enableGradleEnterprise": false,
    "gradleEnterpriseHost": "",
    "applicationType": "monolith",
    "packageName": "com.jhipster.demo.tree",
    "packageFolder": "com/jhipster/demo/tree",
    "jhipsterVersion": "7.5.0",
    "clientPackageManager": "npm",
    "dtoSuffix": "DTO",
    "entitySuffix": "",
    "reactive": false,
    "clientTheme": "none",
    "clientThemeVariant": "",
    "applicationIndex": 0,
    "entities": ["Animal", "Fish", "Bird", "Mammal", "Marsupial", "Rodent", "Cat", "DomesticatedCat"],
    "skipCheckLengthOfIdentifier": false,
    "skipFakeData": false,
    "blueprints": [],
    "otherModules": [],
    "pages": [],
    "nativeLanguage": "en",
    "creationTimestamp": 1645088543533,
    "jwtSecretKey": "YourJWTSecretKeyWasReplacedByThisMeaninglessTextByTheJHipsterInfoCommandForObviousSecurityReasons",    "devServerPort": 4200,
    "lastLiquibaseTimestamp": 1645089023000
  }
}

</pre>
</details>


##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Animal {
  name String required minlength(3)
  description String
}
entity Fish {
  hasScales Boolean required
}
entity Bird {
  wakingBehaviour WakingBehaviour required
}
entity Mammal {
  numberOfLegs Integer required min(2)
  numberOfArms Integer required min(0)
}
entity Marsupial {
  isEndangered Boolean required
}
entity Rodent {
  hasTail Boolean required
}
entity Cat {
  livesLeft Integer required
}
entity DomesticatedCat {
  hairType HairType required
}
enum WakingBehaviour {
  DIURNAL,
  NOCTURNAL
}
enum HairType {
  HAIRLESS,
  LONG,
  SHORT
}

relationship OneToOne {
  Fish{animal required} to Animal
  Bird{animal required} to Animal
  Mammal{animal required} to Animal
  Marsupial{mammal required} to Mammal
  Rodent{mammal required} to Mammal
  Cat{mammal required} to Mammal
  DomesticatedCat{cat required} to Cat
}

paginate Animal, Fish, Bird, Mammal, Marsupial, Rodent with infinite-scroll

</pre>
</details>


##### **Environment and Tools**

openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment Temurin-11.0.13+8 (build 11.0.13+8)
OpenJDK 64-Bit Server VM Temurin-11.0.13+8 (build 11.0.13+8, mixed mode)

git version 2.34.1.windows.1

node: v16.13.2

npm: 8.1.2

Docker version 20.10.12, build e91ed57

Docker Compose version v2.2.3


No change to package.json was detected. No package manager install will be executed.
Congratulations, JHipster execution is complete!


##### **Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<!--
If the error is during an entity creation or associated with a specific entity.
If you are using JDL, please share that configuration as well.
-->

##### **Browsers and Operating System**

<!-- What OS are you on? is this a problem with all browsers or only IE8? -->
Windows 10 Pro
Chrome, but this issue is not browser related

- [x] Checking this box is mandatory (this is just to show you read everything)

<!-- Love JHipster? Please consider supporting our collective:
👉  https://opencollective.com/generator-jhipster/donate -->

--- a/generators/entity-client/templates/common/src/test/javascript/cypress/integration/entity/entity.spec.ts.ejs
+++ b/generators/entity-client/templates/common/src/test/javascript/cypress/integration/entity/entity.spec.ts.ejs
@@ -24,6 +24,7 @@ const otherEntities = _.uniq(Object.values(differentRelationships).filter(rels =
 const skipCreateTest =
   (
     !cypressBootstrapEntities ||
+    requiredRelationships.some(rel => rel.otherEntity.primaryKey && rel.otherEntity.primaryKey.derived) ||
     requiredRelationships.some(rel => rel.otherEntity.builtInUser || rel.otherEntity === this.entity) ||
     requiredRelationships.map(rel => rel.otherEntity.relationships).flat().some(rel => rel.relationshipRequired) ||
     !entityFakeData

--- a/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
+++ b/generators/entity-server/templates/src/test/java/package/web/rest/EntityResourceIT.java.ejs
@@ -517,13 +517,17 @@ if (field.fieldTypeString || field.blobContentTypeText) {
         <%_ } else { _%>
         <%= asEntity(otherEntityNameCapitalized) %> <%= otherEntityName %>;
           <%_ if (databaseTypeSql && !reactive) { _%>
+            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_") { _%>
         if (TestUtil.findAll(em, <%= asEntity(otherEntityNameCapitalized) %>.class).isEmpty()) {
+            <%_ } _%>
             <%= otherEntityName %> = <%= createEntityPrefix %><%= otherEntityNameCapitalized %>ResourceIT.create<% if (fieldStatus === 'UPDATED_') { %>Updated<% } %>Entity(em)<%= createEntityPostfix %>;
             em.persist(<%= otherEntityName %>);
             em.flush();
+            <%_ if (!isUsingMapsId || fieldStatus !== "UPDATED_") { _%>
         } else {
             <%= otherEntityName %> = TestUtil.findAll(em, <%= asEntity(otherEntityNameCapitalized) %>.class).get(0);
         }
+            <%_ } _%>
           <%_ } else { _%>
         <%= otherEntityName %> = <%= createEntityPrefix %><%= otherEntityNameCapitalized %>ResourceIT.create<% if (fieldStatus === 'UPDATED_') { %>Updated<% } %>Entity(<% if (databaseType === 'sql') { %>em<% } %>)<%= createEntityPostfix %>;
           <%_ } _%>

--- a/test-integration/samples/.jhipster/MapsIdChildEntityWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdChildEntityWithDTO.json
@@ -17,6 +17,13 @@
       "ownerSide": true,
       "relationshipName": "mapsIdParentEntityWithDTO",
       "relationshipType": "one-to-one"
+    },
+    {
+      "otherEntityName": "mapsIdGrandchildEntityWithDTO",
+      "otherEntityRelationshipName": "mapsIdChildEntityWithDTO",
+      "ownerSide": false,
+      "relationshipName": "mapsIdGrandchildEntityWithDTO",
+      "relationshipType": "one-to-one"
     }
   ],
   "searchEngine": false,

--- a/test-integration/samples/.jhipster/MapsIdChildEntityWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdChildEntityWithDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "mapstruct",
-  "entityTableName": "maps_id_child_entity_withdto",
+  "entityTableName": "maps_id_child_with_dto",
   "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdChildEntityWithoutDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdChildEntityWithoutDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "no",
-  "entityTableName": "child_entity_wo_dto",
+  "entityTableName": "maps_id_child_wo_dto",
   "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdChildEntityWithoutDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdChildEntityWithoutDTO.json
@@ -17,6 +17,13 @@
       "ownerSide": true,
       "relationshipName": "mapsIdParentEntityWithoutDTO",
       "relationshipType": "one-to-one"
+    },
+    {
+      "otherEntityName": "mapsIdGrandchildEntityWithoutDTO",
+      "otherEntityRelationshipName": "mapsIdChildEntityWithoutDTO",
+      "ownerSide": false,
+      "relationshipName": "mapsIdGrandchildEntityWithoutDTO",
+      "relationshipType": "one-to-one"
     }
   ],
   "searchEngine": false,

--- a/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json
@@ -1,0 +1,24 @@
+{
+  "changelogDate": "20220313234025",
+  "clientRootFolder": "",
+  "databaseType": "sql",
+  "dto": "mapstruct",
+  "entityTableName": "maps_id_gc_entity_w_dto",
+  "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
+  "fluentMethods": true,
+  "jpaMetamodelFiltering": false,
+  "pagination": "no",
+  "relationships": [
+    {
+      "id": true,
+      "otherEntityField": "id",
+      "otherEntityName": "mapsIdChildEntityWithDTO",
+      "otherEntityRelationshipName": "mapsIdGrandchildEntityWithDTO",
+      "ownerSide": true,
+      "relationshipName": "mapsIdChildEntityWithDTO",
+      "relationshipType": "one-to-one"
+    }
+  ],
+  "searchEngine": false,
+  "service": "serviceImpl"
+}

--- a/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "mapstruct",
-  "entityTableName": "maps_id_gc_entity_w_dto",
+  "entityTableName": "maps_id_grandchild_with_dto",
   "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json
@@ -1,0 +1,24 @@
+{
+  "changelogDate": "20220313234416",
+  "clientRootFolder": "",
+  "databaseType": "sql",
+  "dto": "no",
+  "entityTableName": "maps_id_gc_entity_wo_dto",
+  "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
+  "fluentMethods": true,
+  "jpaMetamodelFiltering": false,
+  "pagination": "no",
+  "relationships": [
+    {
+      "id": true,
+      "otherEntityField": "id",
+      "otherEntityName": "mapsIdChildEntityWithoutDTO",
+      "otherEntityRelationshipName": "mapsIdGrandchildEntityWithoutDTO",
+      "ownerSide": true,
+      "relationshipName": "mapsIdChildEntityWithoutDTO",
+      "relationshipType": "one-to-one"
+    }
+  ],
+  "searchEngine": false,
+  "service": "no"
+}

--- a/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdGrandchildEntityWithoutDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "no",
-  "entityTableName": "maps_id_gc_entity_wo_dto",
+  "entityTableName": "maps_id_grandchild_wo_dto",
   "fields": [{ "fieldName": "date", "fieldType": "Instant" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdParentEntityWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdParentEntityWithDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "mapstruct",
-  "entityTableName": "maps_id_parent_entity_withdto",
+  "entityTableName": "maps_id_parent_with_dto",
   "fields": [{ "fieldName": "name", "fieldType": "String" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdParentEntityWithoutDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdParentEntityWithoutDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "no",
-  "entityTableName": "parent_entity_wo_dto",
+  "entityTableName": "maps_id_parent_wo_dto",
   "fields": [{ "fieldName": "name", "fieldType": "String" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/samples/.jhipster/MapsIdUserProfileWithDTO.json
+++ b/test-integration/samples/.jhipster/MapsIdUserProfileWithDTO.json
@@ -3,7 +3,7 @@
   "clientRootFolder": "",
   "databaseType": "sql",
   "dto": "mapstruct",
-  "entityTableName": "maps_id_user_profile_withdto",
+  "entityTableName": "maps_id_user_profile_with_dto",
   "fields": [{ "fieldName": "dateOfBirth", "fieldType": "Instant" }],
   "fluentMethods": true,
   "jpaMetamodelFiltering": false,

--- a/test-integration/scripts/11-generate-config.sh
+++ b/test-integration/scripts/11-generate-config.sh
@@ -150,8 +150,10 @@ elif [[ "$JHI_ENTITY" == "sqlfull" ]]; then
 
     moveEntity MapsIdParentEntityWithoutDTO
     moveEntity MapsIdChildEntityWithoutDTO
+    moveEntity MapsIdGrandchildEntityWithoutDTO
     moveEntity MapsIdParentEntityWithDTO
     moveEntity MapsIdChildEntityWithDTO
+    moveEntity MapsIdGrandchildEntityWithDTO
     moveEntity MapsIdUserProfileWithDTO
 
     moveEntity JpaFilteringRelationship

--- a/test-integration/scripts/11-generate-entities.sh
+++ b/test-integration/scripts/11-generate-entities.sh
@@ -125,8 +125,10 @@ elif [[ "$JHI_ENTITY" == "sqlfull" ]]; then
 
     moveEntity MapsIdParentEntityWithoutDTO
     moveEntity MapsIdChildEntityWithoutDTO
+    moveEntity MapsIdGrandchildEntityWithoutDTO
     moveEntity MapsIdParentEntityWithDTO
     moveEntity MapsIdChildEntityWithDTO
+    moveEntity MapsIdGrandchildEntityWithDTO
     moveEntity MapsIdUserProfileWithDTO
 
     moveEntity JpaFilteringRelationship

--- a/utils/relationship.js
+++ b/utils/relationship.js
@@ -71,7 +71,12 @@ function prepareRelationshipForTemplates(entityWithConfig, relationship, generat
         }
         return otherSideRelationship.relationshipName === relationship.otherEntityRelationshipName;
       });
-      if (
+      if (!otherRelationship) {
+        // TODO throw error at v8.
+        generator.warning(
+          `Error at '${entityName}' definitions: 'otherEntityRelationshipName' is set with value '${relationship.otherEntityRelationshipName}' at relationship '${relationship.relationshipName}' but no back-reference was found at '${otherEntityName}'`
+        );
+      } else if (
         otherRelationship &&
         otherRelationship.otherEntityRelationshipName &&
         otherRelationship.otherEntityRelationshipName !== relationship.relationshipName

